File: backend/geoEvent/src/TicketService/TicketService.Domain/Entities/Reservation.cs
using TicketService.Domain.Enums;
using TicketService.Domain.Exceptions;

namespace TicketService.Domain.Entities;

public class Reservation
{
    public int ReservationId { get; private set; }
    public DateTime ReservedAt { get; private set; }
    public int EventId { get; private set; }
    public int UserId { get; private set; }
    public int? EventTicketId { get; private set; }
    public int Quantity { get; private set; }
    public decimal TotalAmount { get; private set; }
    public string Currency { get; private set; } = "BAM";
    public ReservationStatus Status { get; private set; } = ReservationStatus.Pending;
    public DateTime CreatedAt { get; private set; }
    public DateTime? ConfirmedAt { get; private set; }
    public DateTime? CancelledAt { get; private set; }
    public DateTime? ExpiredAt { get; private set; }
    public DateTime ExpiresAt { get; private set; }
    public string? PaymentReference { get; private set; }
    public string? Notes { get; private set; }

    public string? PendingProviderOrderId { get; private set; }
    public PaymentMethod? PendingPaymentMethod { get; private set; }
    public DateTime? PendingPaymentCreatedAt { get; private set; }

    public EventTicket? EventTicket { get; set; }
    public ICollection<Ticket> Tickets { get; private set; } = [];
    public ICollection<PaymentDetail> PaymentDetails { get; private set; } = [];

    public RefundRequestStatus RefundRequestStatus { get; private set; } = RefundRequestStatus.None;
    public string? RefundReason { get; private set; }
    public DateTime? RefundRequestedAt { get; private set; }
    public DateTime? RefundReviewedAt { get; private set; }
    public int? RefundReviewedByUserId { get; private set; }
    public string? RefundDecisionReason { get; private set; }
    public string? RefundModeratorAction { get; private set; }

    private Reservation() { }

    public static Reservation Create(
        int eventId,
        int userId,
        int eventTicketId,
        int quantity,
        decimal totalAmount,
        string currency,
        DateTime expiresAt,
        string? notes)
    {
        if (eventId <= 0)
            throw new BusinessException("Event ID is required.");

        if (userId <= 0)
            throw new BusinessException("User ID is required.");

        if (eventTicketId <= 0)
            throw new BusinessException("Event ticket ID is required.");

        if (quantity <= 0)
            throw new BusinessException("Quantity must be greater than zero.");

        if (totalAmount < 0)
            throw new BusinessException("Total amount cannot be negative.");

        if (string.IsNullOrWhiteSpace(currency))
            throw new BusinessException("Currency is required.");

        var now = DateTime.UtcNow;

        return new Reservation
        {
            ReservedAt = now,
            EventId = eventId,
            UserId = userId,
            EventTicketId = eventTicketId,
            Quantity = quantity,
            TotalAmount = totalAmount,
            Currency = currency.Trim().ToUpperInvariant(),
            Status = ReservationStatus.Pending,
            CreatedAt = now,
            ExpiresAt = expiresAt,
            Notes = string.IsNullOrWhiteSpace(notes) ? null : notes.Trim()
        };
    }

    public bool IsExpired() =>
        Status == ReservationStatus.Pending && DateTime.UtcNow > ExpiresAt;

    public bool CanBeConfirmed() =>
        Status == ReservationStatus.Pending && !IsExpired();

    public bool CanBeCancelled() =>
        Status == ReservationStatus.Pending || Status == ReservationStatus.Confirmed;

    public void AttachPendingPayment(string providerOrderId, PaymentMethod paymentMethod)
    {
        if (!CanBeConfirmed())
            throw new BusinessException("Reservation cannot accept a pending payment in its current state.");

        if (string.IsNullOrWhiteSpace(providerOrderId))
            throw new BusinessException("Provider order ID is required.");

        PendingProviderOrderId = providerOrderId.Trim();
        PendingPaymentMethod = paymentMethod;
        PendingPaymentCreatedAt = DateTime.UtcNow;
    }

    public void ClearPendingPayment()
    {
        PendingProviderOrderId = null;
        PendingPaymentMethod = null;
        PendingPaymentCreatedAt = null;
    }

    public void Cancel()
    {
        if (!CanBeCancelled())
            throw new BusinessException("Reservation cannot be cancelled in its current state.");

        Status = ReservationStatus.Cancelled;
        CancelledAt = DateTime.UtcNow;
        ClearPendingPayment();
    }

    public void Confirm(string paymentReference)
    {
        if (!CanBeConfirmed())
            throw new BusinessException("Reservation cannot be confirmed in its current state.");

        if (string.IsNullOrWhiteSpace(paymentReference))
            throw new BusinessException("Payment reference is required.");

        Status = ReservationStatus.Confirmed;
        ConfirmedAt = DateTime.UtcNow;
        PaymentReference = paymentReference.Trim();
        ClearPendingPayment();
    }

    public void Expire()
    {
        if (Status != ReservationStatus.Pending)
            throw new BusinessException("Only pending reservations can expire.");

        Status = ReservationStatus.Expired;
        ExpiredAt = DateTime.UtcNow;
        ClearPendingPayment();
    }

    public void Refund()
    {
        if (Status != ReservationStatus.Confirmed)
            throw new BusinessException("Only confirmed reservations can be refunded.");

        Status = ReservationStatus.Refunded;
        CancelledAt = DateTime.UtcNow;
        ClearPendingPayment();
    }

    public bool CanRequestRefund()
    {
        return Status == ReservationStatus.Confirmed &&
               TotalAmount > 0 &&
               RefundRequestStatus != RefundRequestStatus.Pending &&
               RefundRequestStatus != RefundRequestStatus.Approved &&
               RefundRequestStatus != RefundRequestStatus.Processing &&
               RefundRequestStatus != RefundRequestStatus.Refunded;
    }

    public bool CanApproveRefund()
    {
        return RefundRequestStatus == RefundRequestStatus.Pending &&
               Status == ReservationStatus.Confirmed;
    }

    public bool CanRejectRefund()
    {
        return RefundRequestStatus == RefundRequestStatus.Pending &&
               Status == ReservationStatus.Confirmed;
    }

    public bool CanMarkRefundProcessing()
    {
        return (RefundRequestStatus == RefundRequestStatus.Pending ||
                RefundRequestStatus == RefundRequestStatus.Approved) &&
               Status == ReservationStatus.Confirmed;
    }

    public bool CanMarkRefundFailed()
    {
        return RefundRequestStatus == RefundRequestStatus.Processing;
    }

    public bool CanMarkRefundCompleted()
    {
        return (RefundRequestStatus == RefundRequestStatus.Processing ||
                RefundRequestStatus == RefundRequestStatus.Approved) &&
               Status == ReservationStatus.Confirmed;
    }

    public void RequestRefund(string? reason)
    {
        if (!CanRequestRefund())
            throw new BusinessException("Refund cannot be requested for this reservation.");

        RefundRequestStatus = RefundRequestStatus.Pending;
        RefundReason = string.IsNullOrWhiteSpace(reason) ? null : reason.Trim();
        RefundRequestedAt = DateTime.UtcNow;
        RefundReviewedAt = null;
        RefundReviewedByUserId = null;
        RefundDecisionReason = null;
        RefundModeratorAction = null;
    }

    public void MarkRefundApproved(int reviewerUserId, string? decisionReason, string? moderatorAction)
    {
        if (!CanApproveRefund())
            throw new BusinessException("Only pending refund requests can be approved.");

        if (reviewerUserId <= 0)
            throw new BusinessException("Reviewer user ID is required.");

        RefundRequestStatus = RefundRequestStatus.Approved;
        RefundReviewedAt = DateTime.UtcNow;
        RefundReviewedByUserId = reviewerUserId;
        RefundDecisionReason = string.IsNullOrWhiteSpace(decisionReason) ? null : decisionReason.Trim();
        RefundModeratorAction = string.IsNullOrWhiteSpace(moderatorAction) ? null : moderatorAction.Trim();
    }

    public void MarkRefundRejected(int reviewerUserId, string? decisionReason, string? moderatorAction)
    {
        if (!CanRejectRefund())
            throw new BusinessException("Only pending refund requests can be rejected.");

        if (reviewerUserId <= 0)
            throw new BusinessException("Reviewer user ID is required.");

        RefundRequestStatus = RefundRequestStatus.Rejected;
        RefundReviewedAt = DateTime.UtcNow;
        RefundReviewedByUserId = reviewerUserId;
        RefundDecisionReason = string.IsNullOrWhiteSpace(decisionReason) ? null : decisionReason.Trim();
        RefundModeratorAction = string.IsNullOrWhiteSpace(moderatorAction) ? null : moderatorAction.Trim();
    }

    public void MarkRefundProcessing(int reviewerUserId, string? decisionReason, string? moderatorAction)
    {
        if (!CanMarkRefundProcessing())
            throw new BusinessException("Refund request is not ready for processing.");

        if (reviewerUserId <= 0)
            throw new BusinessException("Reviewer user ID is required.");

        RefundRequestStatus = RefundRequestStatus.Processing;
        RefundReviewedAt = DateTime.UtcNow;
        RefundReviewedByUserId = reviewerUserId;
        RefundDecisionReason = string.IsNullOrWhiteSpace(decisionReason) ? null : decisionReason.Trim();
        RefundModeratorAction = string.IsNullOrWhiteSpace(moderatorAction) ? null : moderatorAction.Trim();
    }

    public void MarkRefundFailed(int reviewerUserId, string? decisionReason, string? moderatorAction)
    {
        if (!CanMarkRefundFailed())
            throw new BusinessException("Only processing refund requests can fail.");

        if (reviewerUserId <= 0)
            throw new BusinessException("Reviewer user ID is required.");

        RefundRequestStatus = RefundRequestStatus.Failed;
        RefundReviewedAt = DateTime.UtcNow;
        RefundReviewedByUserId = reviewerUserId;
        RefundDecisionReason = string.IsNullOrWhiteSpace(decisionReason) ? null : decisionReason.Trim();
        RefundModeratorAction = string.IsNullOrWhiteSpace(moderatorAction) ? null : moderatorAction.Trim();
    }

    public void MarkRefundCompleted(int reviewerUserId, string? decisionReason, string? moderatorAction)
    {
        if (!CanMarkRefundCompleted())
            throw new BusinessException("Refund request is not ready to be completed.");

        if (reviewerUserId <= 0)
            throw new BusinessException("Reviewer user ID is required.");

        RefundRequestStatus = RefundRequestStatus.Refunded;
        RefundReviewedAt = DateTime.UtcNow;
        RefundReviewedByUserId = reviewerUserId;
        RefundDecisionReason = string.IsNullOrWhiteSpace(decisionReason) ? null : decisionReason.Trim();
        RefundModeratorAction = string.IsNullOrWhiteSpace(moderatorAction) ? null : moderatorAction.Trim();

        Refund();
    }
}

File: frontend/geo_event/lib/shared/reservations/models/reservation.dart
import '../../tickets/models/request_refund_status.dart';
import 'reservation_status.dart';
import 'ticket.dart';

class Reservation {
  final int reservationId;
  final int userId;
  final int eventId;
  final int? eventTicketId;
  final int quantity;
  final double totalAmount;
  final String currency;
  final String status;
  final DateTime createdAt;
  final DateTime? confirmedAt;
  final DateTime? cancelledAt;
  final DateTime? expiredAt;
  final DateTime expiresAt;
  final String? paymentReference;
  final String? notes;
  final List<Ticket> tickets;

  final String? refundRequestStatus;
  final String? refundReason;
  final DateTime? refundRequestedAt;
  final DateTime? refundReviewedAt;
  final int? refundReviewedByUserId;
  final String? refundDecisionReason;

  const Reservation({
    required this.reservationId,
    required this.userId,
    required this.eventId,
    this.eventTicketId,
    required this.quantity,
    required this.totalAmount,
    required this.currency,
    required this.status,
    required this.createdAt,
    this.confirmedAt,
    this.cancelledAt,
    this.expiredAt,
    required this.expiresAt,
    this.paymentReference,
    this.notes,
    required this.tickets,
    this.refundRequestStatus,
    this.refundReason,
    this.refundRequestedAt,
    this.refundReviewedAt,
    this.refundReviewedByUserId,
    this.refundDecisionReason,
  });

  ReservationStatus get typedStatus {
  final normalized = status.trim().toLowerCase();
  final now = DateTime.now();

  if (normalized == ReservationStatus.cancelled.apiValue.toLowerCase()) {
    return ReservationStatus.cancelled;
  }

  if (normalized == ReservationStatus.refunded.apiValue.toLowerCase()) {
    return ReservationStatus.refunded;
  }

  if (normalized == ReservationStatus.expired.apiValue.toLowerCase()) {
    return ReservationStatus.expired;
  }

  if (normalized == ReservationStatus.confirmed.apiValue.toLowerCase()) {
    return ReservationStatus.confirmed;
  }

  if (normalized == ReservationStatus.pending.apiValue.toLowerCase()) {
    if (expiresAt.isBefore(now) || expiresAt.isAtSameMomentAs(now)) {
      return ReservationStatus.expired;
    }

    return ReservationStatus.pending;
  }

  return ReservationStatus.unknown;
}

String get displayStatus {
  switch (typedStatus) {
    case ReservationStatus.pending:
      return 'Pending';
    case ReservationStatus.confirmed:
      return 'Confirmed';
    case ReservationStatus.cancelled:
      return 'Cancelled';
    case ReservationStatus.expired:
      return 'Expired';
    case ReservationStatus.refunded:
      return 'Refunded';
    case ReservationStatus.unknown:
      return 'Unknown';
  }
}

  bool get isExpiredByTime {
    if (typedStatus != ReservationStatus.pending) {
      return false;
    }

    final now = DateTime.now();
    return expiresAt.isBefore(now) || expiresAt.isAtSameMomentAs(now);
  }

bool get canBeCancelled {
  if (typedStatus == ReservationStatus.pending) return true;
  if (typedStatus == ReservationStatus.confirmed && totalAmount <= 0) return true;
  return false;
}

bool get canBeRefunded {
  return typedStatus == ReservationStatus.confirmed && totalAmount > 0;
}

RefundRequestStatus get typedRefundStatus =>
    RefundRequestStatus.fromValue(refundRequestStatus);

bool get hasPendingRefundRequest =>
    typedRefundStatus == RefundRequestStatus.pending;

bool get isRefundRejected =>
    typedRefundStatus == RefundRequestStatus.rejected;

bool get isRefundCompleted =>
    typedRefundStatus == RefundRequestStatus.refunded ||
    typedStatus == ReservationStatus.refunded;

bool get canRequestRefund {
  return typedStatus == ReservationStatus.confirmed &&
      totalAmount > 0 &&
      typedRefundStatus != RefundRequestStatus.pending &&
      typedRefundStatus != RefundRequestStatus.processing &&
      typedRefundStatus != RefundRequestStatus.approved &&
      typedRefundStatus != RefundRequestStatus.refunded;
}

  bool get isActive {
    return typedStatus != ReservationStatus.cancelled &&
        typedStatus != ReservationStatus.expired &&
        typedStatus != ReservationStatus.refunded;
  }

  factory Reservation.fromJson(Map<String, dynamic> json) {
  DateTime? tryParse(dynamic value) {
    if (value == null) return null;
    return DateTime.tryParse(value.toString())?.toLocal();
  }

  return Reservation(
    reservationId: (json['reservationId'] as num).toInt(),
    userId: (json['userId'] as num).toInt(),
    eventId: (json['eventId'] as num).toInt(),
    eventTicketId: (json['eventTicketId'] as num?)?.toInt(),
    quantity: (json['quantity'] as num).toInt(),
    totalAmount: (json['totalAmount'] as num).toDouble(),
    currency: (json['currency'] ?? '').toString(),
    status: (json['status'] ?? '').toString(),
    createdAt: DateTime.parse((json['createdAt'] ?? '').toString()).toLocal(),
    confirmedAt: tryParse(json['confirmedAt']),
    cancelledAt: tryParse(json['cancelledAt']),
    expiredAt: tryParse(json['expiredAt']),
    expiresAt: DateTime.parse((json['expiresAt'] ?? '').toString()).toLocal(),
    paymentReference: json['paymentReference']?.toString(),
    notes: json['notes']?.toString(),
    tickets: (json['tickets'] as List<dynamic>? ?? const [])
        .whereType<Map>()
        .map((t) => Ticket.fromJson(Map<String, dynamic>.from(t)))
        .toList(),
    refundRequestStatus: json['refundRequestStatus']?.toString(),
    refundReason: json['refundReason']?.toString(),
    refundRequestedAt: tryParse(json['refundRequestedAt']),
    refundReviewedAt: tryParse(json['refundReviewedAt']),
    refundReviewedByUserId: (json['refundReviewedByUserId'] as num?)?.toInt(),
    refundDecisionReason: json['refundDecisionReason']?.toString(),
  );
}

  Reservation copyWith({
  int? reservationId,
  int? userId,
  int? eventId,
  Object? eventTicketId = sentinel,
  int? quantity,
  double? totalAmount,
  String? currency,
  String? status,
  DateTime? createdAt,
  Object? confirmedAt = sentinel,
  Object? cancelledAt = sentinel,
  Object? expiredAt = sentinel,
  DateTime? expiresAt,
  Object? paymentReference = sentinel,
  Object? notes = sentinel,
  List<Ticket>? tickets,
  Object? refundRequestStatus = sentinel,
  Object? refundReason = sentinel,
  Object? refundRequestedAt = sentinel,
  Object? refundReviewedAt = sentinel,
  Object? refundReviewedByUserId = sentinel,
  Object? refundDecisionReason = sentinel,
}) {
  return Reservation(
    reservationId: reservationId ?? this.reservationId,
    userId: userId ?? this.userId,
    eventId: eventId ?? this.eventId,
    eventTicketId: identical(eventTicketId, sentinel)
        ? this.eventTicketId
        : eventTicketId as int?,
    quantity: quantity ?? this.quantity,
    totalAmount: totalAmount ?? this.totalAmount,
    currency: currency ?? this.currency,
    status: status ?? this.status,
    createdAt: createdAt ?? this.createdAt,
    confirmedAt: identical(confirmedAt, sentinel)
        ? this.confirmedAt
        : confirmedAt as DateTime?,
    cancelledAt: identical(cancelledAt, sentinel)
        ? this.cancelledAt
        : cancelledAt as DateTime?,
    expiredAt: identical(expiredAt, sentinel)
        ? this.expiredAt
        : expiredAt as DateTime?,
    expiresAt: expiresAt ?? this.expiresAt,
    paymentReference: identical(paymentReference, sentinel)
        ? this.paymentReference
        : paymentReference as String?,
    notes: identical(notes, sentinel) ? this.notes : notes as String?,
    tickets: tickets ?? this.tickets,
    refundRequestStatus: identical(refundRequestStatus, sentinel)
        ? this.refundRequestStatus
        : refundRequestStatus as String?,
    refundReason: identical(refundReason, sentinel)
        ? this.refundReason
        : refundReason as String?,
    refundRequestedAt: identical(refundRequestedAt, sentinel)
        ? this.refundRequestedAt
        : refundRequestedAt as DateTime?,
    refundReviewedAt: identical(refundReviewedAt, sentinel)
        ? this.refundReviewedAt
        : refundReviewedAt as DateTime?,
    refundReviewedByUserId: identical(refundReviewedByUserId, sentinel)
        ? this.refundReviewedByUserId
        : refundReviewedByUserId as int?,
    refundDecisionReason: identical(refundDecisionReason, sentinel)
        ? this.refundDecisionReason
        : refundDecisionReason as String?,
  );
}

  static const sentinel = Object();
}